File: docs/specs/reference/contentHash.ts
#!/usr/bin/env -S npx tsx
/**
 * Sterish canonical `content_hash` v1 — reference implementation (TypeScript).
 *
 * Normative spec: `docs/specs/content-hash.md`. If this file and the spec ever
 * disagree, the spec wins and this file is the bug.
 *
 * CLI:
 *   npx tsx contentHash.ts <dir>              # hash a skill directory, print 64 hex chars
 *   npx tsx contentHash.ts --vectors [path]   # run the shared test vectors, print report lines
 *
 * The `--vectors` report is byte-for-byte comparable with the Python and Rust
 * reference implementations; `scripts/verify-content-hash.sh` diffs them.
 */

import { createHash } from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Domain-separation prefix. 24 bytes, trailing newline included. */
export const MAGIC: Uint8Array = new TextEncoder().encode("sterish-content-hash/v1\n");
if (MAGIC.length !== 24) throw new Error("MAGIC must be exactly 24 bytes");

const EXCLUDED_DIRS = new Set([".git", "node_modules", "__pycache__", ".venv", "target"]);
const EXCLUDED_FILES = new Set([".DS_Store"]);
const EXCLUDED_SUFFIXES = [".pyc"];

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

export type ErrorKind = "EmptyFileSet" | "DuplicatePath" | "InvalidPath" | "NotUtf8";

export class ContentHashError extends Error {
  constructor(public readonly kind: ErrorKind, message: string) {
    super(`${kind}: ${message}`);
    this.name = "ContentHashError";
  }
}

// ---------------------------------------------------------------------------
// Primitives
// ---------------------------------------------------------------------------

export interface SkillFile {
  /** UTF-8 bytes of the skill-root-relative POSIX path. */
  pathBytes: Uint8Array;
  /** Raw file bytes, BEFORE normalization. */
  raw: Uint8Array;
}

/** Unsigned 32-bit big-endian length prefix. */
export function u32be(n: number): Uint8Array {
  if (!Number.isInteger(n) || n < 0 || n > 0xffff_ffff) {
    throw new Error(`value out of u32 range: ${n}`);
  }
  return new Uint8Array([(n >>> 24) & 0xff, (n >>> 16) & 0xff, (n >>> 8) & 0xff, n & 0xff]);
}

/**
 * ASC bytewise comparison on RAW bytes.
 * Deliberately NOT `String.prototype.localeCompare` and NOT the default
 * `Array.prototype.sort()` on strings — that is UTF-16 code-unit order, which
 * disagrees with UTF-8 byte order for non-BMP code points (see the
 * `non-bmp-path-order` vector).
 */
export function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    if (a[i] !== b[i]) return a[i] < b[i] ? -1 : 1;
  }
  return a.length === b.length ? 0 : a.length < b.length ? -1 : 1;
}

const STRICT_UTF8 = new TextDecoder("utf-8", { fatal: true });

function isUtf8(bytes: Uint8Array): boolean {
  try {
    STRICT_UTF8.decode(bytes);
    return true;
  } catch {
    return false;
  }
}

export function checkPath(pathBytes: Uint8Array): void {
  if (pathBytes.length === 0) throw new ContentHashError("InvalidPath", "empty path");
  if (!isUtf8(pathBytes)) throw new ContentHashError("InvalidPath", "path is not valid UTF-8");
  const text = STRICT_UTF8.decode(pathBytes);
  if (text.includes("\\")) {
    throw new ContentHashError("InvalidPath", `backslash is not a path separator: ${text}`);
  }
  if (text.includes("\0")) {
    throw new ContentHashError("InvalidPath", `NUL byte in path: ${text}`);
  }
  for (const part of text.split("/")) {
    if (part === "") {
      throw new ContentHashError(
        "InvalidPath",
        `empty path component (leading/trailing/double slash): ${text}`,
      );
    }
    if (part === ".") throw new ContentHashError("InvalidPath", `'.' component not allowed: ${text}`);
    if (part === "..") throw new ContentHashError("InvalidPath", `'..' component not allowed: ${text}`);
  }
}

/**
 * Step (a) every CRLF -> LF, step (b) every remaining CR -> LF, step (c) strip
 * ALL trailing LF. Written as three literal passes so it maps 1:1 onto the
 * spec text; do not "optimize" it into a single pass without re-running
 * scripts/verify-content-hash.sh.
 */
export function normalizeContent(raw: Uint8Array): Uint8Array {
  if (!isUtf8(raw)) throw new ContentHashError("NotUtf8", "content is not valid UTF-8");

  const CR = 0x0d;
  const LF = 0x0a;

  // (a) CRLF -> LF (leftmost, non-overlapping).
  const stepA = new Uint8Array(raw.length);
  let n = 0;
  for (let i = 0; i < raw.length; ) {
    if (raw[i] === CR && i + 1 < raw.length && raw[i + 1] === LF) {
      stepA[n++] = LF;
      i += 2;
    } else {
      stepA[n++] = raw[i];
      i += 1;
    }
  }

  // (b) remaining CR -> LF.
  for (let i = 0; i < n; i++) {
    if (stepA[i] === CR) stepA[i] = LF;
  }

  // (c) strip all trailing LF.
  let end = n;
  while (end > 0 && stepA[end - 1] === LF) end--;
  return stepA.subarray(0, end);
}

export function canonicalBytes(files: SkillFile[]): Uint8Array {
  if (files.length === 0) {
    throw new ContentHashError("EmptyFileSet", "a skill must contain at least one file");
  }

  const seen = new Set<string>();
  const items: Array<{ pathBytes: Uint8Array; content: Uint8Array }> = [];
  for (const f of files) {
    checkPath(f.pathBytes);
    const key = Array.from(f.pathBytes).join(",");
    if (seen.has(key)) {
      throw new ContentHashError("DuplicatePath", `duplicate path: ${STRICT_UTF8.decode(f.pathBytes)}`);
    }
    seen.add(key);
    items.push({ pathBytes: f.pathBytes, content: normalizeContent(f.raw) });
  }

  items.sort((a, b) => compareBytes(a.pathBytes, b.pathBytes));

  const chunks: Uint8Array[] = [MAGIC, u32be(items.length)];
  for (const item of items) {
    chunks.push(u32be(item.pathBytes.length), item.pathBytes);
    chunks.push(u32be(item.content.length), item.content);
  }
  const total = chunks.reduce((acc, c) => acc + c.length, 0);
  const buf = new Uint8Array(total);
  let off = 0;
  for (const c of chunks) {
    buf.set(c, off);
    off += c.length;
  }
  return buf;
}

/** 64 lowercase hex chars. */
export function contentHash(files: SkillFile[]): string {
  return createHash("sha256").update(canonicalBytes(files)).digest("hex");
}

// ---------------------------------------------------------------------------
// Directory packager
// ---------------------------------------------------------------------------

export function collectDir(root: string): SkillFile[] {
  const encoder = new TextEncoder();
  const out: SkillFile[] = [];
  const walk = (dir: string, relParts: string[]): void => {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true }).sort((a, b) =>
      a.name < b.name ? -1 : a.name > b.name ? 1 : 0,
    )) {
      if (entry.isDirectory()) {
        if (EXCLUDED_DIRS.has(entry.name)) continue;
        walk(path.join(dir, entry.name), [...relParts, entry.name]);
      } else if (entry.isFile()) {
        if (EXCLUDED_FILES.has(entry.name)) continue;
        if (EXCLUDED_SUFFIXES.some((s) => entry.name.endsWith(s))) continue;
        const rel = [...relParts, entry.name].join("/");
        out.push({
          pathBytes: encoder.encode(rel),
          raw: new Uint8Array(fs.readFileSync(path.join(dir, entry.name))),
        });
      }
    }
  };
  walk(path.resolve(root), []);
  return out;
}

export function hashDir(root: string): string {
  return contentHash(collectDir(root));
}

// ---------------------------------------------------------------------------
// Test-vector runner (shared report format)
// ---------------------------------------------------------------------------

interface VectorFile {
  path: string;
  content_b64: string;
}
interface Vector {
  id: string;
  files: VectorFile[];
  expected_sha256?: string;
  expect_equal_to?: string[];
  expect_differs_from?: string[];
}
interface ErrorCase {
  id: string;
  files: VectorFile[];
  expect_error: ErrorKind;
}
interface VectorDoc {
  vectors: Vector[];
  error_cases: ErrorCase[];
}

function filesOf(v: { files: VectorFile[] }): SkillFile[] {
  const encoder = new TextEncoder();
  return v.files.map((f) => ({
    pathBytes: encoder.encode(f.path),
    raw: new Uint8Array(Buffer.from(f.content_b64, "base64")),
  }));
}

function runVectors(vectorPath: string): number {
  const doc: VectorDoc = JSON.parse(fs.readFileSync(vectorPath, "utf8"));
  const lines: string[] = [];
  const problems: string[] = [];
  const hashes = new Map<string, string>();

  for (const v of doc.vectors) {
    const got = contentHash(filesOf(v));
    hashes.set(v.id, got);
    lines.push(`VECTOR ${v.id} ${got}`);
    if (v.expected_sha256 && v.expected_sha256 !== got) {
      problems.push(`${v.id}: expected ${v.expected_sha256}, computed ${got}`);
    }
  }

  for (const v of doc.vectors) {
    for (const other of v.expect_equal_to ?? []) {
      const ok = hashes.get(v.id) === hashes.get(other);
      lines.push(`RELATION ${v.id} equals ${other} ${ok ? "OK" : "FAIL"}`);
      if (!ok) problems.push(`${v.id} must equal ${other} but does not`);
    }
    for (const other of v.expect_differs_from ?? []) {
      const ok = hashes.get(v.id) !== hashes.get(other);
      lines.push(`RELATION ${v.id} differs ${other} ${ok ? "OK" : "FAIL"}`);
      if (!ok) problems.push(`${v.id} must differ from ${other} but does not`);
    }
  }

  for (const c of doc.error_cases) {
    let gotKind = "NO_ERROR";
    try {
      contentHash(filesOf(c));
    } catch (e) {
      gotKind = e instanceof ContentHashError ? e.kind : `UNEXPECTED(${String(e)})`;
    }
    lines.push(`ERROR ${c.id} ${gotKind}`);
    if (gotKind !== c.expect_error) {
      problems.push(`${c.id}: expected error ${c.expect_error}, got ${gotKind}`);
    }
  }

  process.stdout.write(lines.join("\n") + "\n");
  for (const p of problems) process.stderr.write(`typescript: ${p}\n`);
  return problems.length > 0 ? 1 : 0;
}

const HERE = path.dirname(fileURLToPath(import.meta.url));
const DEFAULT_VECTORS = path.join(HERE, "..", "vectors", "content-hash-vectors.json");

function main(argv: string[]): number {
  if (argv[0] === "--vectors") return runVectors(argv[1] ?? DEFAULT_VECTORS);
  if (argv.length !== 1 || argv[0] === "-h" || argv[0] === "--help") {
    process.stderr.write("usage: contentHash.ts <dir> | --vectors [vectors.json]\n");
    return 2;
  }
  process.stdout.write(hashDir(argv[0]) + "\n");
  return 0;
}

if (process.argv[1] && fileURLToPath(import.meta.url) === path.resolve(process.argv[1])) {
  process.exit(main(process.argv.slice(2)));
}

File: docs/specs/reference/content_hash.py
#!/usr/bin/env python3
"""Sterish canonical ``content_hash`` v1 — reference implementation (Python).

Normative spec: ``docs/specs/content-hash.md``. If this file and the spec ever
disagree, the spec wins and this file is the bug.

CLI
---
    python3 content_hash.py <dir>              # hash a skill directory, print 64 hex chars
    python3 content_hash.py --vectors [path]   # run the shared test vectors, print report lines
    python3 content_hash.py --regen [path]     # recompute expected_sha256 in the vector file

The ``--vectors`` report is byte-for-byte comparable with the TypeScript and
Rust reference implementations; ``scripts/verify-content-hash.sh`` diffs them.
"""

from __future__ import annotations

import base64
import hashlib
import json
import os
import struct
import sys
from pathlib import Path
from typing import Iterable, List, Sequence, Tuple

# --------------------------------------------------------------------------
# Constants
# --------------------------------------------------------------------------

#: Domain-separation prefix. 24 bytes, trailing newline included.
MAGIC = b"sterish-content-hash/v1\n"
assert len(MAGIC) == 24, "MAGIC must be exactly 24 bytes"

#: Directories dropped by the packager BEFORE hashing (not part of the hash).
EXCLUDED_DIRS = frozenset({".git", "node_modules", "__pycache__", ".venv", "target"})
#: File names dropped by the packager BEFORE hashing.
EXCLUDED_FILES = frozenset({".DS_Store"})
#: File suffixes dropped by the packager BEFORE hashing.
EXCLUDED_SUFFIXES = (".pyc",)


# --------------------------------------------------------------------------
# Errors
# --------------------------------------------------------------------------


class ContentHashError(Exception):
    """Base class. ``kind`` is the stable, cross-language error name."""

    kind = "ContentHashError"


class EmptyFileSet(ContentHashError):
    kind = "EmptyFileSet"


class DuplicatePath(ContentHashError):
    kind = "DuplicatePath"


class InvalidPath(ContentHashError):
    kind = "InvalidPath"


class NotUtf8(ContentHashError):
    kind = "NotUtf8"


# --------------------------------------------------------------------------
# Primitives
# --------------------------------------------------------------------------


def u32be(n: int) -> bytes:
    """Unsigned 32-bit big-endian length prefix."""
    if n < 0 or n > 0xFFFF_FFFF:
        raise ContentHashError(f"value out of u32 range: {n}")
    return struct.pack(">I", n)


def check_path(path_bytes: bytes) -> None:
    """Reject anything that is not a clean, relative, POSIX path."""
    if not path_bytes:
        raise InvalidPath("empty path")
    try:
        text = path_bytes.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise InvalidPath("path is not valid UTF-8") from exc
    if "\\" in text:
        raise InvalidPath(f"backslash is not a path separator: {text!r}")
    if "\x00" in text:
        raise InvalidPath(f"NUL byte in path: {text!r}")
    for part in text.split("/"):
        if part == "":
            raise InvalidPath(f"empty path component (leading/trailing/double slash): {text!r}")
        if part == ".":
            raise InvalidPath(f"'.' component not allowed: {text!r}")
        if part == "..":
            raise InvalidPath(f"'..' component not allowed: {text!r}")


def normalize_content(raw: bytes) -> bytes:
    """Apply the v1 content normalization: CRLF -> LF, CR -> LF, strip trailing LF."""
    try:
        raw.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise NotUtf8(f"content is not valid UTF-8: {exc}") from exc
    normalized = raw.replace(b"\r\n", b"\n").replace(b"\r", b"\n")
    return normalized.rstrip(b"\n")


def canonical_bytes(files: Sequence[Tuple[bytes, bytes]]) -> bytes:
    """Build CANON from ``(path_bytes, raw_content)`` pairs. Order of input is irrelevant."""
    if not files:
        raise EmptyFileSet("a skill must contain at least one file")

    seen: set[bytes] = set()
    items: List[Tuple[bytes, bytes]] = []
    for path_bytes, raw in files:
        check_path(path_bytes)
        if path_bytes in seen:
            raise DuplicatePath(f"duplicate path: {path_bytes!r}")
        seen.add(path_bytes)
        items.append((path_bytes, normalize_content(raw)))

    # ASC bytewise on the RAW path bytes. Python's bytes ordering is bytewise.
    items.sort(key=lambda item: item[0])

    buf = bytearray(MAGIC)
    buf += u32be(len(items))
    for path_bytes, content in items:
        buf += u32be(len(path_bytes))
        buf += path_bytes
        buf += u32be(len(content))
        buf += content
    return bytes(buf)


def content_hash(files: Sequence[Tuple[bytes, bytes]]) -> str:
    """64 lowercase hex chars."""
    return hashlib.sha256(canonical_bytes(files)).hexdigest()


# --------------------------------------------------------------------------
# Directory packager
# --------------------------------------------------------------------------


def is_excluded(rel_parts: Sequence[str], name: str) -> bool:
    if any(part in EXCLUDED_DIRS for part in rel_parts):
        return True
    if name in EXCLUDED_FILES:
        return True
    return name.endswith(EXCLUDED_SUFFIXES)


def collect_dir(root: str | os.PathLike[str]) -> List[Tuple[bytes, bytes]]:
    """Walk a skill directory, applying the exclusion list, and return hashable files."""
    root_path = Path(root).resolve()
    out: List[Tuple[bytes, bytes]] = []
    for dirpath, dirnames, filenames in os.walk(root_path):
        dirnames[:] = sorted(d for d in dirnames if d not in EXCLUDED_DIRS)
        rel_dir = Path(dirpath).relative_to(root_path)
        rel_parts = [] if rel_dir == Path(".") else list(rel_dir.parts)
        for name in sorted(filenames):
            if is_excluded(rel_parts, name):
                continue
            rel = "/".join([*rel_parts, name])
            out.append((rel.encode("utf-8"), Path(dirpath, name).read_bytes()))
    return out


def hash_dir(root: str | os.PathLike[str]) -> str:
    return content_hash(collect_dir(root))


# --------------------------------------------------------------------------
# Test-vector runner (shared report format)
# --------------------------------------------------------------------------

DEFAULT_VECTORS = Path(__file__).resolve().parent.parent / "vectors" / "content-hash-vectors.json"


def _files_of(vector: dict) -> List[Tuple[bytes, bytes]]:
    return [
        (f["path"].encode("utf-8"), base64.b64decode(f["content_b64"], validate=True))
        for f in vector["files"]
    ]


def run_vectors(path: Path) -> int:
    doc = json.loads(path.read_text(encoding="utf-8"))
    lines: List[str] = []
    problems: List[str] = []
    hashes: dict[str, str] = {}

    for vector in doc["vectors"]:
        vid = vector["id"]
        got = content_hash(_files_of(vector))
        hashes[vid] = got
        lines.append(f"VECTOR {vid} {got}")
        expected = vector.get("expected_sha256")
        if expected and expected != got:
            problems.append(f"{vid}: expected {expected}, computed {got}")

    for vector in doc["vectors"]:
        vid = vector["id"]
        for other in vector.get("expect_equal_to", []):
            ok = hashes[vid] == hashes[other]
            lines.append(f"RELATION {vid} equals {other} {'OK' if ok else 'FAIL'}")
            if not ok:
                problems.append(f"{vid} must equal {other} but does not")
        for other in vector.get("expect_differs_from", []):
            ok = hashes[vid] != hashes[other]
            lines.append(f"RELATION {vid} differs {other} {'OK' if ok else 'FAIL'}")
            if not ok:
                problems.append(f"{vid} must differ from {other} but does not")

    for case in doc["error_cases"]:
        cid = case["id"]
        want = case["expect_error"]
        try:
            content_hash(_files_of(case))
            got_kind = "NO_ERROR"
        except ContentHashError as exc:
            got_kind = exc.kind
        lines.append(f"ERROR {cid} {got_kind}")
        if got_kind != want:
            problems.append(f"{cid}: expected error {want}, got {got_kind}")

    print("\n".join(lines))
    for problem in problems:
        print(f"python: {problem}", file=sys.stderr)
    return 1 if problems else 0


def regen_vectors(path: Path) -> int:
    doc = json.loads(path.read_text(encoding="utf-8"))
    for vector in doc["vectors"]:
        vector["expected_sha256"] = content_hash(_files_of(vector))
    path.write_text(json.dumps(doc, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    print(f"regenerated {len(doc['vectors'])} expected_sha256 values in {path}", file=sys.stderr)
    return 0


def main(argv: Sequence[str]) -> int:
    if len(argv) >= 1 and argv[0] == "--vectors":
        return run_vectors(Path(argv[1]) if len(argv) > 1 else DEFAULT_VECTORS)
    if len(argv) >= 1 and argv[0] == "--regen":
        return regen_vectors(Path(argv[1]) if len(argv) > 1 else DEFAULT_VECTORS)
    if len(argv) != 1 or argv[0] in ("-h", "--help"):
        print(__doc__, file=sys.stderr)
        return 2
    print(hash_dir(argv[0]))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
